working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,559 @@
+        "description": "New Description"
+    }), content_type='application/json')
+
+    assert response.status_code == 400
+    assert b"Title is required" in response.data
+
+def test_update_book_without_description(client):
+    response = client.put('/books/1', data=json.dumps({
+        "title": "New Title"
+    }), content_type='application/json')
+
+    assert response.status_code == 400
+    assert b"Description is required" in response.data
+```
+
+We could manually update the tests to match better, but Copilot can also reformat them to be closer to what we're looking for if we follow up with a prompt like:
+
+> Please update the tests above to match the formatting used in test_routes.py
+
+The resulting test code Copilot suggests still has issues we will address, but it is structured much closer to our other tests:
+
+```
+def test_update_book_with_extra_keys(client, two_saved_books):
+    # Arrange
+    book_id = 1
+    data = {
+        "title": "Updated Book",
+        "description": "New description",
+        "extra_key": "Extra Value"
+    }
+
+    # Act
+    response = client.put(f"/books/{book_id}", json=data)
+    response_body = response.get_json()
+
+    # Assert
+    assert response.status_code == 200
+    assert response_body == "Book #1 successfully updated"
+
+def test_update_book_without_title(client, two_saved_books):
+    # Arrange
+    book_id = 1
+    data = {
+        "description": "New description"
+    }
+
+    # Act
+    response = client.put(f"/books/{book_id}", json=data)
+
+    # Assert
+    assert response.status_code == 400
+    assert b"Title is required" in response.data
+
+def test_update_book_without_description(client, two_saved_books):
+    # Arrange
+    book_id = 1
+    data = {
+        "title": "Updated Book"
+    }
+
+    # Act
+    response = client.put(f"/books/{book_id}", json=data)
+
+    # Assert
+    assert response.status_code == 400
+    assert b"Description is required" in response.data
+```
+
+If we paste these tests into our test file and run them, the test `test_update_book_with_extra_keys` passes without changes, but we need to take a deeper look at the tests for missing required data. 
+
+The assertions for these test are assuming that a 400 status code will be sent in case of error and made an assumption about what the error message in the response would be. If we navigate over to the `update_book` function in `routes.py`, we don't actually have any error handling for this scenario. These tests fail because our application crashes with a `KeyError` when required data is missing from the request dictionary. 
+
+Copilot could have written tests that expected a crash, but here it has helped us by highlighting error handling that would be useful to have. Rather than changing the tests to expect a crash, we're going to add error handling to `update_book` to make our app more robust, then update our test assertions to check for the correct message.
```

Comment:
- rephrased to make it sound less like copilot considered the alternative and rejected it
- reminder of what a crash looks like to the client
```suggestion
While it's possible that Copilot could have written tests that expected a crash (which would look like a `500 Internal Server Error`), it's much more common that these scenarios would result in a `400 Bad Request`. So even though Copilot generated failing tests, it turns out that this helped us by highlighting error handling that would be useful to have. Rather than changing the tests to expect a crash, we're going to add error handling to `update_book` to make our app more robust, then update our test assertions to check for the correct message.
```